jerry-core/ecma/base/ecma-helpers-string.c

```diff
@@ -216,6 +216,120 @@ ecma_new_ecma_string_from_utf8 (const lit_utf8_byte_t *string_p, /**< utf-8 stri
 } /* ecma_new_ecma_string_from_utf8 */
 
 /**
+ * Allocate a new ecma-string and initialize it from the utf8 string argument.
+ * All 4-bytes long unicode sequences are converted into two 3-bytes long sequences.
+ *
+ * @return pointer to ecma-string descriptor
+ */
+ecma_string_t *
+ecma_new_ecma_string_from_utf8_converted_to_cesu8 (const lit_utf8_byte_t *string_p, /**< utf-8 string */
+                                                   lit_utf8_size_t string_size) /**< utf-8 string size */
+{
+  JERRY_ASSERT (string_p != NULL || string_size == 0);
+
+  ecma_string_t *string_desc_p = NULL;
+
+  ecma_length_t str_length = 0;
+  lit_utf8_size_t conv_size = 0;
+  lit_utf8_size_t size = 0;
+
+  /* Calculate the required length and size information of the converted cesu-8 encoded string */
+  while (size < string_size)
+  {
+    if ((string_p[size] & LIT_UTF8_1_BYTE_MASK) == LIT_UTF8_1_BYTE_MARKER)
+    {
+      size++;
+    }
+    else if ((string_p[size] & LIT_UTF8_2_BYTE_MASK) == LIT_UTF8_2_BYTE_MARKER)
+    {
+      size += 2;
+    }
+    else if ((string_p[size] & LIT_UTF8_3_BYTE_MASK) == LIT_UTF8_3_BYTE_MARKER)
+    {
+      size += 3;
+    }
+    else
+    {
+      JERRY_ASSERT ((string_p[size] & LIT_UTF8_4_BYTE_MASK) == LIT_UTF8_4_BYTE_MARKER);
+      size += 4;
+      conv_size += 2;
+    }
+
+    str_length++;
+  }
+
+  JERRY_ASSERT (size == string_size);
```

Comment:
If this is true, there is no need to computing it except for debugging purposes. This would simplify the for loop above.